packages/govuk-frontend/src/govuk/all.mjs

```diff
@@ -29,76 +29,39 @@ function initAll(config) {
     return
   }
 
+  const components = /** @type {const} */ ([
+    [Accordion, config.accordion],
+    [Button, config.button],
+    [CharacterCount, config.characterCount],
+    [Checkboxes],
+    [ErrorSummary, config.errorSummary],
+    [ExitThisPage, config.exitThisPage],
+    [Header],
+    [NotificationBanner, config.notificationBanner],
+    [Radios],
+    [SkipLink],
+    [Tabs]
+  ])
+
   // Allow the user to initialise GOV.UK Frontend in only certain sections of the page
   // Defaults to the entire document if nothing is set.
   const $scope = config.scope instanceof HTMLElement ? config.scope : document
 
-  const $accordions = $scope.querySelectorAll('[data-module="govuk-accordion"]')
-  $accordions.forEach(($accordion) => {
-    new Accordion($accordion, config.accordion)
-  })
-
-  const $buttons = $scope.querySelectorAll('[data-module="govuk-button"]')
-  $buttons.forEach(($button) => {
-    new Button($button, config.button)
-  })
-
-  const $characterCounts = $scope.querySelectorAll(
-    '[data-module="govuk-character-count"]'
-  )
-  $characterCounts.forEach(($characterCount) => {
-    new CharacterCount($characterCount, config.characterCount)
-  })
-
-  const $checkboxes = $scope.querySelectorAll(
-    '[data-module="govuk-checkboxes"]'
-  )
-  $checkboxes.forEach(($checkbox) => {
-    new Checkboxes($checkbox)
-  })
-
-  // Find first error summary module to enhance.
-  const $errorSummary = $scope.querySelector(
-    '[data-module="govuk-error-summary"]'
-  )
-  if ($errorSummary) {
-    new ErrorSummary($errorSummary, config.errorSummary)
-  }
-
-  const $exitThisPageButtons = $scope.querySelectorAll(
-    '[data-module="govuk-exit-this-page"]'
-  )
-  $exitThisPageButtons.forEach(($button) => {
-    new ExitThisPage($button, config.exitThisPage)
-  })
-
-  // Find first header module to enhance.
-  const $header = $scope.querySelector('[data-module="govuk-header"]')
-  if ($header) {
-    new Header($header)
-  }
-
-  const $notificationBanners = $scope.querySelectorAll(
-    '[data-module="govuk-notification-banner"]'
-  )
-  $notificationBanners.forEach(($notificationBanner) => {
-    new NotificationBanner($notificationBanner, config.notificationBanner)
-  })
-
-  const $radios = $scope.querySelectorAll('[data-module="govuk-radios"]')
-  $radios.forEach(($radio) => {
-    new Radios($radio)
-  })
-
-  // Find first skip link module to enhance.
-  const $skipLink = $scope.querySelector('[data-module="govuk-skip-link"]')
-  if ($skipLink) {
-    new SkipLink($skipLink)
-  }
-
-  const $tabs = $scope.querySelectorAll('[data-module="govuk-tabs"]')
-  $tabs.forEach(($tabs) => {
-    new Tabs($tabs)
+  components.forEach(([Component, config]) => {
+    const $elements = $scope.querySelectorAll(
+      `[data-module="${Component.moduleName}"]`
+    )
+
+    $elements.forEach(($element) => {
+      try {
+        // Only pass config to components that accept it
+        'defaults' in Component
```

Comment:
**suggestion** This line should probably rely on the presence of `config` now it's here rather than some internals of `Component`?

Reply:
Give it a try if you like? You'll see that it's flagged `@private` 😉

But every component with static `defaults` _also_ has a config

Reply:
I think @romaricpascal was suggesting checking whether the local `config` variable was defined, rather than looking at `Component.config`?

I've tried switching this out for 

```js
typeof config !== 'undefined'
```

But the TypeScript compiler doesn't like it:

```
../govuk-frontend/src/govuk/all.mjs:60:37 - error TS2554: Expected 1 arguments, but got 2.

60           ? new Component($element, config)
                                       ~~~~~~
```

Reply:
Yeah, that too

Adding another local config to `components` doesn't mean `Component.config` exists

But for `'defaults' in Component` the compiler knows "all components with defaults also have a config"

Reply:
Otherwise we could change that to `'config' in Component` but we'd need to remove `@private`

Reply:
> I think @romaricpascal was suggesting checking whether the local config variable was defined, rather than looking at Component.config?

Yup, that's what I was suggesting, to rely on something local rather than mix responsibilities for `defaults` (providing a default configuration *and* being a marker of the availability of a config from the array of components)  😊

Reply:
Yeah, would be nice

Sadly it's _our_ marker and doesn't tell the compiler anything about the component

E.g. We could add 3x fake configs but doesn't mean they exist on the component

```js
[CharacterCount, config.fake123],
[Checkboxes, config.fake123],
[ErrorSummary, config.fake123],
```

Reply:
I think I prefer the `defaults` check as it's that much simpler. To avoid blocking the PR whilst @romaricpascal is off, going to merge this for now but happy to revisit later.